Command/ListClientsCommand.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace Trikoder\Bundle\OAuth2Bundle\Command;
+
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Trikoder\Bundle\OAuth2Bundle\Manager\ClientFilter;
+use Trikoder\Bundle\OAuth2Bundle\Manager\ClientManagerInterface;
+use Trikoder\Bundle\OAuth2Bundle\Model\Client;
+use Trikoder\Bundle\OAuth2Bundle\Model\Grant;
+use Trikoder\Bundle\OAuth2Bundle\Model\RedirectUri;
+use Trikoder\Bundle\OAuth2Bundle\Model\Scope;
+
+final class ListClientsCommand extends Command
+{
+    public const ALLOWED_COLUMNS = ['identifier', 'secret', 'scope', 'redirect uri', 'grant type'];
+
+    protected static $defaultName = 'trikoder:oauth2:list-clients';
+    private $clientManager;
+
+    public function __construct(ClientManagerInterface $clientManager)
+    {
+        parent::__construct();
+        $this->clientManager = $clientManager;
+    }
+
+    protected function configure(): void
+    {
+        $this
+            ->setDescription('Lists existing oAuth2 clients')
+            ->addOption(
+                'columns',
+                null,
+                InputOption::VALUE_IS_ARRAY | InputOption::VALUE_REQUIRED,
+                'Determine which columns are shown. Can be used multiple times to specify multiple columns.',
+                self::ALLOWED_COLUMNS
+            )
+            ->addOption(
+                'redirect-uri',
+                null,
+                InputOption::VALUE_IS_ARRAY | InputOption::VALUE_REQUIRED,
+                'Finds by redirect uri for client. Use this option multiple times to filter by multiple redirect URIs.',
+                []
+            )
+            ->addOption(
+                'grant-type',
+                null,
+                InputOption::VALUE_IS_ARRAY | InputOption::VALUE_REQUIRED,
+                'Finds by allowed grant type for client. Use this option multiple times to filter by multiple grant types.',
+                []
+            )
+            ->addOption(
+                'scope',
+                null,
+                InputOption::VALUE_IS_ARRAY | InputOption::VALUE_OPTIONAL,
```

Comment:
@Allypost `InputOption::VALUE_OPTIONAL`?

Reply:
@HypeMC The one that got away :(